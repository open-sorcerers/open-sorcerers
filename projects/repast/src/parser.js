import {
  mergeRight,
  objOf,
  all,
  ap,
  concat,
  curry,
  equals,
  filter,
  ifElse,
  indexOf,
  init,
  last,
  map,
  nth,
  of,
  pathOr,
  pipe,
  prop,
  reduce,
  split,
  splitWhen,
  startsWith
} from "ramda"
import { parse } from "recast"
import { trace } from "xtrace"
import { C, L, K, REGEXES, TYPES } from "./constants"

const box = x => (Array.isArray(x) ? x : [x])

const readNestedSignatures = (agg, xx) => {
  const starter = xx.startsWith(C.openParens)
  const ender = xx.endsWith(C.closeParens)
  const { nested, value, lastWasEnder } = agg
  const sliced = { start: init(value), end: last(value) }

  const yy = xx
    .replace(REGEXES.startsWithParens, C.empty)
    .replace(REGEXES.endsWithParens, C.empty)
  return {
    lastWasEnder: ender,
    value: starter
      ? concat(value, box(yy))
      : nested
      ? concat(box(init(value)), [concat(box(last(value)), box(yy))])
      : concat(value, [yy]),
    nested: ender ? false : nested || starter ? true : nested
  }
}

const j2 = x => JSON.stringify(x, null, 2)
const hasString = curry((a, b) => indexOf(a, b) > -1)
const trim = x => x.trim()
const hindleymilnerize = pipe(
  split(C.space),
  map(trim),
  map(z => (z[0] === C.asterisk ? z.slice(1, Infinity) : z)),
  filter(Boolean),
  reduce(readNestedSignatures, { nested: false, value: [] }),
  prop(L.value),
  ifElse(
    x => x[0] === K.HM_TYPE,
    pipe(nth(1), objOf(L.name), mergeRight({ type: TYPES.HM_TYPE })),
    ifElse(
      pipe(
        of,
        ap([pipe(nth(0), equals(K.HM)), pipe(nth(2), equals(C.doubleColon))]),
        all(x => x)
      ),
      x => ({
        [L.type]: TYPES.HM_SIGNATURE,
        [L.name]: nth(1, x),
        [L.signature]: x.slice(3, Infinity)
      }),
      trace("This is not a correctly formatted comment?")
    )
  )
)

// @hm parseWithConfig :: Object -> String -> Array
export const parseWithConfig = curry((config, str) =>
  pipe(
    parse,
    pathOr([], [L.program, L.body]),
    reduce((agg, expression) => {
      const { comments } = expression
      if (
        comments.length &&
        filter(pipe(prop(L.value), hasString(K.HM)), comments).length
      ) {
        return pipe(
          map(prop(L.value)),
          filter(hasString(K.HM)),
          map(comment =>
            mergeRight(hindleymilnerize(comment), { ast: expression })
          ),
          concat(agg)
        )(comments)
      }
      return agg
    }, [])
  )(str)
)
