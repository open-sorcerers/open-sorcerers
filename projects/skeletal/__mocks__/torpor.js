const path = require("path")
const R = require("ramda")
const torpor = jest.genMockFromModule("torpor")
const F = require("fluture")

let mockFiles = Object.create(null)
function __setMockFiles(neu) {
  mockFiles = Object.create(null)
  for (const file in neu) {
    const dir = path.dirname(file)
    if (!mockFiles[dir]) mockFiles[dir] = []
    mockFiles[dir].push(path.basename(file))
  }
}
function readFile(file, config) {
  return new F.Future((bad, good) => {
    try {
      setTimeout(
        () => (file.includes("fail") ? bad(false) : good(config.raw || "")),
        1
      )
      return () => {}
    } catch (e) {
      bad(e)
      return () => {}
    }
  })
}
function writeFile(filepath, content, config) {
  return new F.Future((bad, good) => {
    try {
      setTimeout(() => (filepath.includes("fail") ? bad(false) : good(true)), 1)
      return () => {}
    } catch (e) {
      bad(e)
      return () => {}
    }
  })
}

torpor.__setMockFiles = __setMockFiles
torpor.writeFile = R.curryN(3, writeFile)
torpor.readFile = R.curryN(2, readFile)

module.exports = torpor
