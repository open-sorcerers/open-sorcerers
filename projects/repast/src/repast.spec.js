import { readFile } from "torpor"
import { __, pipe, map, prop } from "ramda"
import { fork } from "fluture"
import { parseWithConfig } from "./parser"
import { parse } from "recast"

test("basic", done => {
  pipe(
    readFile(__, "utf8"),
    map(parseWithConfig({})),
    fork(done)(outcome => {
      expect(map(prop("signature"), outcome)).toMatchSnapshot()
      done()
    })
  )("./fixture/example.js")
})
