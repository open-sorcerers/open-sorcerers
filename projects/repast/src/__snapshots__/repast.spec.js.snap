// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic 1`] = `
Array [
  Array [
    "Number",
    "->",
    "String",
    "->",
    "Number",
  ],
  Array [
    "Number",
    "->",
    "String",
    "->",
    "Number",
  ],
  Array [
    "Number",
    "->",
    "String",
    "->",
    "Number",
  ],
  Array [
    "Number",
    "->",
    "String",
    "->",
    "Number",
  ],
  undefined,
  Array [
    "x",
    "->",
    "Container",
    "x",
  ],
  Array [
    Array [
      "a",
      "->",
      "b",
    ],
    "->",
    "Container",
    "a",
    "->",
    "Container",
    "b",
  ],
]
`;
